daml-lf/spec/daml-lf-1.rst

```diff
@@ -2192,20 +2150,11 @@ types that satisfies the following rules::
   ——————————————————————————————————————————————————— TypeOrderTypeRepUpdate
     'TypeRep' <ₜ 'GenMap'
 
-  ——————————————————————————————————————————————————— TypeOrderGenMapAny
+  ——————————————————————————————————————————————————— TypeOrderGenMapAnyException
     'GenMap' <ₜ 'AnyException'
 
-  ——————————————————————————————————————————————————— TypeOrderAnyExceptionGeneralError
-    'AnyException' <ₜ 'GeneralError'
-
-  ——————————————————————————————————————————————————— TypeOrderGeneralErrorArithmeticError
-    'GeneralError' <ₜ 'ArithmeticError'
-
-  ——————————————————————————————————————————————————— TypeOrderArithmeticErrorContractError
-    'ArithmeticError' <ₜ 'ContractError'
-
-  ——————————————————————————————————————————————————— TypeOrderContractErrorBigNumeric
-    'ContractError' <ₜ 'BigNumeric'
+  ——————————————————————————————————————————————————— TypeOrderAnyExceptionBigNumeric
+    'AnyException' <ₜ 'BigNumeric'
```

Comment:
@remyhaemmerle-da
Are AnyException values comparable (in the value order)?
Is `AnyException` ordered as a type (in the type order)?

Reply:
https://github.com/digital-asset/daml/blob/fb6f72c81d983eebe91b87f491238e4d414573c2/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TypeOrdering.scala#L92 for the type order. Afaict, it’s not orderable by value atm see https://github.com/digital-asset/daml/blob/main/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala but I think it should be. I don’t see a compelling reason why `Any` is orderable but `AnyException` is not.

Reply:
We have a minor issue here.
Either we make it comparable, and we have to revive `ArithmeticError` either we wait we have the LF standard library :cry:

Reply:
We only need it in the spec though not exposed right?

Reply:
As soon you make ArithmeticError (wrapped in a AnyException) and an user Exception (wrapped in a AnyException) comparable and fix the order, I consider it as exposed.

Reply:
In other word, once you make it comparable, you have to keep it in the speedy code.

Reply:
Ah I see because moving it to a standard library would change the order and unless we go for a magic package id (like 0000 or ffff…) we have no way to make sure that ArithmeticError comes always before or always after everything else. In that case, I think we’re better off not adding ordering for now.

Reply:
👍 ok, so AnyException values are not comparable for now.


Also I just noticed / found that `AnyException` as a type is ordered after `RoundingMode` (in the TypeRep order), so I'll update the spec.

Reply:
We cannot really make `AnyException` comparable in the future, if we do not do it now.  
Comparison is not version aware, and I think it is pretty impossible to make it version aware.